src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/parser/Parser.java

```diff
@@ -0,0 +1,501 @@
+                    var type = attributes.get("type", Attribute.Valued.class);
+                    String typeValue = type.isPresent() ? type.get().value() : "link";
+                    if (!typeValue.equals("link") && !typeValue.equals("linkplain")) {
+                        throw new ParseException("Unknown link type: '%s'".formatted(typeValue), t.lineSourceOffset + t.markupLineOffset + type.get().valueStartPosition());
+                    }
+                    Annotate a = new Annotate(new Style.Link(target.value()),
+                                              createRegexPattern(substring, regex, ".+", t.lineSourceOffset + t.markupLineOffset), // different regex not to include newline
+                                              text.subText(t.start(), t.end()));
+                    actions.add(a);
+                }
+                case "replace" -> {
+                    var replacement = attributes.get("replacement", Attribute.Valued.class)
+                            .orElseThrow(() -> new ParseException("replacement is absent", t.lineSourceOffset + t.markupLineOffset + t.nameLineOffset));
+                    Replace a = new Replace(replacement.value(),
+                                            createRegexPattern(substring, regex, t.lineSourceOffset + t.markupLineOffset),
+                                            text.subText(t.start(), t.end()));
+                    actions.add(a);
+                }
+                case "highlight" -> {
+                    var type = attributes.get("type", Attribute.Valued.class);
+
+                    String typeValue = type.isPresent() ? type.get().value() : "bold";
+
+                    Annotate a = new Annotate(new Style.Name(typeValue),
+                                              createRegexPattern(substring, regex, t.lineSourceOffset + t.markupLineOffset),
+                                              text.subText(t.start(), t.end()));
+                    actions.add(a);
+                }
+                case "start" -> {
+                    var region = attributes.get("region", Attribute.Valued.class)
+                            .orElseThrow(() -> new ParseException("Unnamed start", t.lineSourceOffset + t.markupLineOffset + t.nameLineOffset));
+                    String regionValue = region.value();
+                    if (regionValue.isBlank()) {
+                        throw new ParseException("Blank region name", t.lineSourceOffset + t.markupLineOffset + region.valueStartPosition());
+                    }
+                    if (t.attributes().size() != 1) {
+                        throw new ParseException("Unexpected attributes", t.lineSourceOffset + t.markupLineOffset + t.nameLineOffset);
+                    }
+                    actions.add(new Bookmark(region.value(), text.subText(t.start(), t.end() - 1)));
```

Comment:
This method is a bit hard on the eyes and runs quite wide. Maybe add a few more line breaks on those very long lines?

Reply:
So just to be clear, Hannes. Do you suggest to structure the rules like this?

```
pre.snippet {
    background-color: #e8e8e8;
    padding: 10px;
    margin: 12px 0;
    overflow: auto;
    white-space: pre;
}

pre.snippet .italic {
    font-style: italic;
}

pre.snippet .bold {
    font-weight: bold;
}

pre.snippet .highlighted {
    background-color: sandybrown;
    border-radius: 10%;
}
```

Reply:
Yes, that's what I would propose. It limits the definitions to contents of `<pre class="snippet">` elements. Another idea would have been to use snippet-specific class names such as `.snippet-italic`, but that's more verbose and less elegant.